### ISSUE TITLE
Question regarding the Amazon Request Quota/Rate ?

### ISSUE THREAD

**Issue by iosoft** [CPIGroup/phpAmazonMWS#164]
Hello,

This is not a issue, rather a question.
I am using your Library for my PHP project.

Amazon has Request Quota/Rate for almost each and every kind the requests.
Is your Library handling the quota/rate internally? or I have to manage it from my side of coding?

Please let me know.
Thanks and regards.

**Comment by Peardian** [comment]
Hi, good question. The direct answer to your question is "yes and no", as throttling within the Amazon library is a bit tricky.

In the beginning, this library was designed to manage all of the aspects of the request rate, which is why there is an `environment.php` file filled with variables representing the different rate limits and restore rates. However, this proved to be very difficult to design for, so we opted for a much simpler system. If the library receives a 503 error (request rate exceeded), the library will sleep for an amount of time corresponding to the restore rate of the request and try again. There is a setting method, `setThrottleStop()`, that can be used to suppress the sleep/retry.

Then, Amazon went and added an extra hourly request limit. This complicates things further, because it means that sleeping for the usual amount of time will not allow the request to continue like normal, and the library won't be able to tell why the request is being throttled.

There is a third type of throttling that applies to product requests. These requests behave differently in that Amazon will limit the amount of data returned, rather than outright rejecting the request. The library currently cannot detect or stop this kind of throttling.

So, to summarize, the library will be able to handle any typical throttling that occurs, but it is still a good idea to be aware of the API's limitations when designing your code. If you design your script to send many more requests than what Amazon allows, your code will take much longer to finish than if the requests were spaced out.

I hope that answers your question.

**Comment by iosoft** [comment]
This is a great explanation.
You should make a wiki on this point.

Actually, I not not so expert like you, but when I was going through your library, I saw some throttling code. Now its clear.

And also, **thanks for quick and detailed reply.**

**Comment by Peardian** [comment]
Thanks, glad I could help.

Also, you bring up a good idea. When I find the time, I will add a section to the readme file about throttling.
